alvr/sockets/src/lib.rs

```diff
@@ -93,3 +96,180 @@ fn set_dscp(socket: &Socket, dscp: Option<DscpTos>) {
+        stream_config_packet: impl Serialize,
+        socket_config: StreamSocketConfig,
+    ) -> ConResult<Self> {
+        let client_ip = control_socket.inner.peer_addr().to_con()?.ip();
+
+        control_socket.send(&stream_config_packet).to_con()?;
+
+        control_socket
+            .send(&ServerControlPacket::StartStream)
+            .to_con()?;
+
+        let signal = control_socket.recv(timeout)?;
+        if !matches!(signal, ClientControlPacket::StreamReady) {
+            con_bail!("Got unexpected packet waiting for stream ack");
+        }
+
+        let stream_socket = StreamSocketBuilder::connect_to_client(
+            timeout,
+            client_ip,
+            socket_config.port,
+            socket_config.protocol,
+            socket_config.dscp,
+            socket_config.send_buffer_bytes,
+            socket_config.recv_buffer_bytes,
+            socket_config.max_packet_size,
+        )?;
+
+        Ok(Self {
+            control_socket,
+            stream_socket,
+        })
+    }
+
+    // Must be called by the client
+    pub fn from_server_connection(
+        mut control_socket: ProtoControlSocket,
+        timeout: Duration,
+        socket_config: StreamSocketConfig,
+    ) -> ConResult<ServerConnectionResult> {
+        let server_ip = control_socket.inner.peer_addr().to_con()?.ip();
+
+        match control_socket.recv(timeout)? {
+            ServerControlPacket::StartStream => (),
+            ServerControlPacket::Restarting => return Ok(ServerConnectionResult::Restarting),
+            _ => con_bail!("Got unexpected packet waiting for stream start"),
+        }
+
+        let stream_socket_builder = StreamSocketBuilder::listen_for_server(
+            timeout,
+            socket_config.port,
+            socket_config.protocol,
+            socket_config.dscp,
+            socket_config.send_buffer_bytes,
+            socket_config.recv_buffer_bytes,
+        )
+        .to_con()?;
+
+        control_socket
+            .send(&ClientControlPacket::StreamReady)
+            .to_con()?;
+
+        let stream_socket = stream_socket_builder.accept_from_server(
+            server_ip,
+            socket_config.port,
+            socket_config.max_packet_size,
+            timeout,
+        )?;
+
+        Ok(ServerConnectionResult::Connected(Self {
+            control_socket,
+            stream_socket,
+        }))
+    }
+
+    pub fn request_reliable_stream<T>(&self) -> ConResult<ControlSocketSender<T>> {
+        Ok(ControlSocketSender {
+            inner: self.control_socket.inner.try_clone().to_con()?,
+            buffer: vec![],
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn subscribe_to_reliable_stream<T>(&self) -> ConResult<ControlSocketReceiver<T>> {
+        Ok(ControlSocketReceiver {
+            inner: self.control_socket.inner.try_clone().to_con()?,
+            buffer: vec![],
+            recv_state: None,
+            _phantom: PhantomData,
+        })
+    }
```

Comment:
I think you should rather try to figure out how to put this in the other file, because making everything `pub(crate)` for these two functions is really bad practice

Reply:
Or like make a helper there to create these or something

Reply:
This is just (1) of multiple PRs. This PR focuses on just creating a wrapper over the existing stream and control socket, (2) will decompose and recreate a unified codebase of the SocketConnection. (3) will be a breaking change that doesn't duplicate the TCP socket between control and stream sockets